python/ray/util/sgd/torch/worker_group.py

```diff
@@ -0,0 +1,432 @@
+
+    def _check_potential_remote_workers_size(self):
+        # ASSUME 1 GPU + 1 CPU is already reserved for the local worker
+        remote_resources = ray.available_resources()
+        max_remote_workers = self.num_workers
+        new_remote_workers = min(
+            remote_resources.get("CPU", 0), max_remote_workers)
+        if self.use_gpu:
+            new_remote_workers = min(
+                remote_resources.get("GPU", 0), new_remote_workers)
+        return new_remote_workers
+
+    def resize_workers(self, max_retries=10):
+        self._reset()
+
+        time.sleep(1)
+        for i in range(max_retries):
+            new_remote_workers = self._check_potential_remote_workers_size()
+            if new_remote_workers:
+                self._last_resize = time.time()
+                self.start_workers(int(new_remote_workers))
+                self.load_state_dict(self.state_dict())
+                return
+            else:
+                delay = 2**i
+                logger.warning(
+                    "No new workers found. Retrying in %d sec." % delay)
+                time.sleep(delay)
+        raise RuntimeError("Exceeded max_retries for relaunching workers.")
+
+    def get_num_workers(self):
+        return len(self.remote_workers)
+
+
+class LocalWorkerGroup(BaseWorkerGroup):
+    def __init__(self):
+        self.local_worker = DeactivatedRunner()
+        self.remote_worker_group = RemoteWorkerGroup()
+        self.num_workers = 0
+
+    def start_workers(self, num_workers, params, initialization_hook,
+                      timeout_s, num_cpus_per_worker, use_gpu):
+        logger.debug(f"start_workers: Setting %d workers." % num_workers)
+
+        self.num_workers = num_workers
+
+        if num_workers == 1:
+            self.local_worker = TorchRunner(**params)
+            if initialization_hook:
+                self.apply_all_workers(initialization_hook)
+            self.local_worker.setup()
+        else:
+            # Start local worker
+            self.local_worker = LocalDistributedRunner(
+                num_cpus=num_cpus_per_worker, num_gpus=int(use_gpu), **params)
+            self.remote_worker_group._init_workers(
+                num_workers - 1, params, num_cpus_per_worker, use_gpu)
+            if initialization_hook:
+                self.apply_all_workers(initialization_hook)
+
+            # Compute URL for initializing distributed PyTorch
+            address = setup_address()
+
+            remote_components = self.remote_worker_group._setup_components()
+            self.local_worker.setup_components()
+            ray.get(remote_components)
+
+            remote_pgs = self.remote_worker_group._setup_process_group(
+                address, timeout_s, num_workers)
+            self.local_worker.setup_process_group(address, num_workers - 1,
+                                                  num_workers,
+                                                  timedelta(timeout_s))
```

Comment:
does this uniquely assign rank to all workers? (i think it does but just checking)

Reply:
Yes it should- the remote runners have rank [0, num_workers-2] and the local runner has rank num_workers-1.